compiler/rustc_middle/src/hir/map/mod.rs

```diff
@@ -987,7 +977,15 @@ pub(super) fn crate_hash(tcx: TyCtxt<'_>, crate_num: CrateNum) -> Svh {
     source_file_names.sort_unstable();
 
     let mut stable_hasher = StableHasher::new();
-    node_hashes.hash_stable(&mut hcx, &mut stable_hasher);
+    for (def_path_hash, fingerprint, def_id) in hir_body_nodes.iter() {
+        def_path_hash.0.hash_stable(&mut hcx, &mut stable_hasher);
+        fingerprint.hash_stable(&mut hcx, &mut stable_hasher);
+        AttributeMap { map: &tcx.untracked_crate.attrs, prefix: *def_id }
+            .hash_stable(&mut hcx, &mut stable_hasher);
+        let span = tcx.definitions.def_span(*def_id);
+        debug_assert_eq!(span.parent(), None);
+        span.hash_stable(&mut hcx, &mut stable_hasher);
+    }
```

Comment:
I don't understand what this commit does.

Reply:
Metadata encoding only serializes "absolute" spans. As a consequence, those "absolute" spans must appear in the crate hash. Equivalently, the reference spans with respect to which the "relative" spans  are produced must appear in the crate hash. The reference spans are the `definitions.def_span` (this name is suboptimal, `source_span` is more consistent with the query).